[material-ui][Autocomplete] Fix listbox opens and closes on click when used with `limitTags`

appleSimple opened the pull request:
Fixes #42432

~~Input element is fall down because of hide tag was shown.
So event target not same in the code below.
I fixed it by add style property `pointer-events: none;` at autocomplete input. To attach [mdn link](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events#none) for you.~~

```
// Autocomplete.js

onClick: (event) => {
  if (event.target === event.currentTarget) {
    handleInputMouseDown(event);
  }
},
```

See discussion below in the pull request.

Before: https://mui.com/material-ui/react-autocomplete/#limit-tags
After: https://deploy-preview-42494--material-ui.netlify.app/material-ui/react-autocomplete/#limit-tags

ZeeshanTamboli reviewed (changes requested):
This works as shown in the documentation:

Before: https://mui.com/material-ui/react-autocomplete/#limit-tags
After: https://deploy-preview-42494--material-ui.netlify.app/material-ui/react-autocomplete/#limit-tags

However, the tests fail as shown in the CI.

appleSimple commented:
Thank you for read. @ZeeshanTamboli .

I passed CI test done! 🎉 please check this result.

Additionally, I deleted previous fixes and pushed another. Because of cause was wrong I expected.
When I clicked Input, occured events that Input's mouse down and wrapper's click so the list box is open and close because of double events.

### Input area
![image](https://github.com/mui/material-ui/assets/73057489/2ae3b2a4-ed7e-4b5d-bfda-aea3929ac9c8)

### Wrapper area
![image](https://github.com/mui/material-ui/assets/73057489/a82b3edf-7445-4dab-86a4-557cb76a75d2)

ZeeshanTamboli reviewed (commented):
This seem to work well. But there is a slight delay in the tag getting rendered after the listbox is opened as shown in the video:

https://github.com/mui/material-ui/assets/20900032/375978fa-9fc8-4fb6-96c9-0efb851835ba

Could that be avoided?

appleSimple commented:
hi, @ZeeshanTamboli 

I think the same way about slightly delay movement.
I tried several things and finally i fixed it done. The cause remains the same.

https://github.com/mui/material-ui/assets/73057489/6586d478-98d4-4e6b-ba06-af3c2728a6d4


check result please 🙏

ZeeshanTamboli reviewed (commented):
The current solution works but is not ideal. The bug is a regression from #36369. The following fix resolves the issue:

```diff
--- a/packages/mui-base/src/useAutocomplete/useAutocomplete.js
+++ b/packages/mui-base/src/useAutocomplete/useAutocomplete.js
@@ -1042,6 +1042,7 @@ export function useAutocomplete(props) {
   const handleInputMouseDown = (event) => {
     if (!disabledProp && (inputValue === '' || !open)) {
       handlePopupIndicator(event);
+      event.stopPropagation();
     }
   };

diff --git a/packages/mui-material/src/Autocomplete/Autocomplete.js b/packages/mui-material/src/Autocomplete/Autocomplete.js
index c8ba035f39..8dc63ca14e 100644
--- a/packages/mui-material/src/Autocomplete/Autocomplete.js
+++ b/packages/mui-material/src/Autocomplete/Autocomplete.js
@@ -724,11 +724,7 @@ const Autocomplete = React.forwardRef(function Autocomplete(inProps, ref) {
             ref: setAnchorEl,
             className: classes.inputRoot,
             startAdornment,
-            onClick: (event) => {
-              if (event.target === event.currentTarget) {
-                handleInputMouseDown(event);
-              }
-            },
+            onMouseDown: (event) => handleInputMouseDown(event),
             ...((hasClearIcon || hasPopupIcon) && {
               endAdornment: (
                 <AutocompleteEndAdornment className={classes.endAdornment} ownerState={ownerState}>
diff --git a/packages/mui-material/src/Autocomplete/Autocomplete.test.js b/packages/mui-material/src/Autocomplete/Autocomplete.test.js
index cc2550cb23..3a3e831e9a 100644
--- a/packages/mui-material/src/Autocomplete/Autocomplete.test.js
+++ b/packages/mui-material/src/Autocomplete/Autocomplete.test.js
@@ -1109,7 +1109,7 @@ describe('<Autocomplete />', () => {
         />,
       );

-      fireEvent.click(ref.current);
+      fireEvent.mouseDown(ref.current);
       expect(handleOpen.callCount).to.equal(1);
     });
```

### Explanation:

The issue was that the `mousedown` event on the input (`handleInputMouseDown` handler) was being called twice, leading to the initial opening and then immediate closing of the listbox. The `open` state was toggled in `handleInputMouseDown` because it was executed twice. This happened because the `mousedown` event from the [input's onMouseDown](https://github.com/mui/material-ui/blob/next/packages/mui-material/src/Autocomplete/Autocomplete.js#L769) bubbled up to the [root's onClick](https://github.com/mui/material-ui/blob/next/packages/mui-material/src/Autocomplete/Autocomplete.js#L727), causing `handleInputMouseDown` [here](https://github.com/mui/material-ui/blob/next/packages/mui-material/src/Autocomplete/Autocomplete.js#L729) to be executed again. 

By the time the `click` event is fired, the event has already propagated through the `mousedown` and `mouseup` phases, potentially causing the `target` and `currentTarget` to be the same if the event bubbled up to the parent div.

The solution above stops the event from bubbling from the input's `mousedown` event and changes the root div's `onClick` to `onMouseDown` because we are handling the input's `mousedown` event. This approach also retains the fix from #36369.

Additionally, we need to add a test case for this with the `limitTags` prop to prevent future regressions.

appleSimple commented:
I didn't know if it was okay to modify test code.
Thanks for the explanation 😊

ZeeshanTamboli commented:
@appleSimple Could you add a test case for this using the `limitTags` prop? The listbox should open but not close immediately on click. The test should pass here but fail in the `next` branch, ensuring it tests the correct behavior.

appleSimple commented:
@ZeeshanTamboli 
My answer was too late, and I will try to make test case you said.
I guessed cause of fail in the `next` branch is "when clicked the arrow". (like below video)
This changed was occured same issue.

I will fix it, too.

https://github.com/mui/material-ui/assets/73057489/e064ea89-2d7d-4d93-b517-76675385f1bd

ZeeshanTamboli commented on packages/mui-base/src/useAutocomplete/useAutocomplete.js:
After merging #42907 to remove the dependency on `@mui/base`, we need to update the [Material UI `useAutocomplete` file](https://github.com/mui/material-ui/blob/next/packages/mui-material/src/useAutocomplete/useAutocomplete.js) which is now used by the Autocomplete component. Can you apply this logic there instead?

appleSimple replied:
I'm done this job!

ZeeshanTamboli commented on packages/mui-base/src/useAutocomplete/useAutocomplete.js:
```diff
@@ -1130,6 +1131,7 @@ export function useAutocomplete(props) {
       tabIndex: -1,
       type: 'button',
       onClick: handlePopupIndicator,
+      onMouseDown: (event) => event.stopPropagation(),
```
Why this change?

appleSimple replied:
When the indicator was clicked, the `mouseDown event` of the root props triggered, causing the `handleOpen` function to execute. Subsequently, the intended indicator event occurred, resulting in the `handleClose` function being triggered.
To address this, I prevented the [propagation of the `mouseDown event` on the indicator](https://github.com/mui/material-ui/blob/8eef03cb64432f7e7b96c5b5459db8ee8f0da448/packages/mui-material/src/useAutocomplete/useAutocomplete.js#L1134).

appleSimple replied:
bug is [this comment](https://github.com/mui/material-ui/pull/42494#issuecomment-2216405861)

appleSimple commented:
I have added the test code. Could you please check if it was added as you intended?

Due to a recent PR that removed dependencies, the files that need to be updated have changed. Therefore, I have added the necessary code to those files. ([#42907](https://github.com/mui/material-ui/pull/42907))

Additionally, ESLint tests are failing due to a TypeScript version issue, and I'm not sure how to resolve this.

ZeeshanTamboli commented on packages/mui-material/src/Autocomplete/Autocomplete.test.js:
```diff
@@ -413,6 +413,29 @@ describe('<Autocomplete />', () => {
       }
     });
+
+    it('listbox stays open after opening', async () => {
```
This test passes in the `next` HEAD branch, so it isn't testing correctly what's needed. It should pass in this branch but fail in the `next` branch.

appleSimple replied:
I got it, I'll modify it, soon!!

ZeeshanTamboli commented on packages/mui-material/src/Autocomplete/Autocomplete.test.js:
```diff
@@ -413,6 +413,29 @@ describe('<Autocomplete />', () => {
       }
     });
+
+    it('listbox stays open after opening', async () => {
```
```suggestion
    it('listbox stays open after opening', () => {
```

appleSimple commented:
@ZeeshanTamboli 
I clearly included the cause of the bug in the test title. I wrote the test code forgetting this. Sorry.
Please check this change.

ZeeshanTamboli commented on packages/mui-material/src/Autocomplete/Autocomplete.test.js:
```diff
@@ -413,6 +413,40 @@ describe('<Autocomplete />', () => {
       }
     });
+
+    it('When the input box needs to expand downward, the listbox should remain open.', () => {
```
This is also passing on the `next` branch. It should fail there. `handleClose` should be called once there.

appleSimple commented:
@ZeeshanTamboli
I tried several tests on the 'next' branch. please check it!

ZeeshanTamboli commented on packages/mui-material/src/Autocomplete/Autocomplete.test.js:
```diff
@@ -413,6 +413,46 @@ describe('<Autocomplete />', () => {
+        />,
+      );
+
+      fireEvent.mouseDown(ref.current);
```
`mousedown` on `ref.current` means clicking the TextField's border, not the `input`. The `handleClose` call count should be once in the `next` branch, so this test isn't accurate. It's difficult to test this. I'll try other methods.

appleSimple replied:
The `input` was implemented with a `click event` in the `next branch`, but it has been changed to a `mouseDown` event in this branch. Will the function test code, rather than DOM tests, yield valid results in both branches? I tried various codes to simulate `mouseDown` the `TextField`, but both branches continued to show the same results, and I wasn't sure what to do next. I really want to solve this problem. I would appreciate an explanation after the modifications.

ZeeshanTamboli replied:
I tried multiple ways. It is hard to simulate the behavior of the bug in the `next` branch in test to make the test fail on `next` and pass here.

appleSimple replied:
okay, I will closed this PR. Thank you!

appleSimple replied:
I thought it was happening in [PR](https://github.com/mui/material-ui/issues/42432), so I closed PR. sorry.

Can I change the way I fixed the bug and try writing the test code again? I'm not sure, but it might be a good idea to try another method.

ZeeshanTamboli replied:
Feel free to.

ZeeshanTamboli reviewed (approved):
@appleSimple Thanks for the pull request.
